url27a7ae3508_serkan_ozal_ocean_of_memories_tgz-pJ8-com_zeroturnaround_rebellabs_oceanofmemories_article1_objectlayout_ObjectMemoryLayoutDemoJ8/src/com/zeroturnaround/rebellabs/oceanofmemories/common/util/JvmUtil.java

```diff
@@ -0,0 +1,2125 @@
+            final Field[] fields = c.getDeclaredFields();
+            for (final Field f : fields) {
+                if (!Modifier.isStatic(f.getModifiers())) {
+                    shallowInstanceSize = adjustForField(shallowInstanceSize, f);
+
+                    if (!f.getType().isPrimitive()) {
+                        f.setAccessible(true);
+                        referenceFields.add(f);
+                    }
+                }
+            }
+        }
+
+        long size = shallowSizeOfInstance(clazz);
+        Object array = Array.newInstance(clazz, 0);
+        int arrayBaseOffset = unsafe.arrayBaseOffset(array.getClass());
+        int arrayIndexScale = unsafe.arrayIndexScale(array.getClass());
+        long classAddress = addressOfClassInternal(clazz);
+        cachedInfo = new ClassInfo(alignObjectSize(shallowInstanceSize),
+                referenceFields.toArray(new Field[referenceFields.size()]),
+                size, arrayBaseOffset, arrayIndexScale, classAddress);
+        return cachedInfo;
+    }
+
+    private static ClassInfo getClassInfo(final Class<?> clazz) {
+        ClassInfo cacheEntry = classCache.get(clazz);
+        if (cacheEntry == null) {
+            cacheEntry = createClassInfo(clazz);
+            classCache.put(clazz, cacheEntry);
+        }
+        return cacheEntry;
+    }
+
+    private static long adjustForField(long sizeSoFar, final Field f) {
+        f.setAccessible(true);
+        final Class<?> type = f.getType();
+        final int fsize = sizeOfType(type);
+        long offsetPlusSize = 0;
+        if (Modifier.isStatic(f.getModifiers())) {
+            offsetPlusSize = unsafe.staticFieldOffset(f) + fsize;
+        } else {
+            offsetPlusSize = unsafe.objectFieldOffset(f) + fsize;
+        }
+        return Math.max(sizeSoFar, offsetPlusSize);
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <T> T allocateInstance(Class<T> clazz) {
+        try {
+            return (T) unsafe.allocateInstance(clazz);
+        } catch (InstantiationException e) {
+            logger.error("Unable to instantiate class: " + clazz.getName(), e);
+            return null;
+        }
+    }
+
+    public static String toHexAddress(long address) {
+        return "0x" + Long.toHexString(address).toUpperCase();
+    }
+
+    public static String toBinaryStringAddress(long address) {
+        return "0x" + Long.toBinaryString(address).toUpperCase();
+    }
+
+    private static void findProcessId() throws Exception {
+        RuntimeMXBean mxbean = ManagementFactory.getRuntimeMXBean();
+        Field jvmField = mxbean.getClass().getDeclaredField("jvm");
+
+        jvmField.setAccessible(true);
+        VMManagement management = (VMManagement) jvmField.get(mxbean);
+        Method method = management.getClass().getDeclaredMethod("getProcessId");
+        method.setAccessible(true);
+        processId = (Integer) method.invoke(management);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Catch `InvocationTargetException` explicitly when you call `Method.invoke()`. Use `Throwable.getCause()` or `getTargetException()` in the catch handler to access more information about the underlying exception. See the following diffs on GitHub that make this code change: [diff1](https://github.com/EsfingeFramework/gamification/commit/20792047e6c69457cca03e877f2ada578188f32a#diff-6808c12fe9ddaea042db84f95dd5ba78adab99d0cb18bb4d85916a8723530768L21), [diff2](https://github.com/cucumber-attic/cuke4duke/commit/f4821d400cf2590126f1298a8dff9f7a33d45dc5#diff-80adaffc20c0e6e7d2e42312386ffe1ddc10213bf5b8e8919895fd85d666c8deL34)

Learn more: [InvocationTargetException in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/InvocationTargetException.html)